aarch64 / arm64 support

I see failure in running [kprobe example](https://github.com/cilium/ebpf/tree/master/examples/kprobe) for aarch64 kernel version `4.19.113`

```
GOOS=linux GOARCH=arm64 go build -o ebpf-kprobe

./ebpf-kprobe 
2021/03/25 22:03:40 error while loading objects: field KprobeExecve: program kprobe_execve: load program without BTF: invalid argument
```

looks like `BPF_PROG_LOAD` is [failing](https://gist.github.com/shrkamat/d24a81505833391f95cbe81f2c865c1a#file-epbf-kprobe-fail-L406)

Finally had the chance to look at this. Have ALARM set up on a rpi4, and this is the test output: https://gist.github.com/ti-mo/033e7de02d58da9abc6a2bac91458291

Everything seems to work for the most part, but it seems like the following map/prog types just don't work on this kernel build:
```
--- FAIL: TestHaveMapType (0.10s)
    --- FAIL: TestHaveMapType/XSKMap (0.00s)
        map_test.go:64: Map type XSKMap isn't supported even though kernel is at least 4.18: not supported
    --- FAIL: TestHaveMapType/InodeStorage (0.00s)
        map_test.go:64: Map type InodeStorage isn't supported even though kernel is at least 5.10: not supported
    --- FAIL: TestHaveMapType/TaskStorage (0.00s)
        map_test.go:64: Map type TaskStorage isn't supported even though kernel is at least 5.11: not supported
--- FAIL: TestHaveProgType (0.00s)
    --- FAIL: TestHaveProgType/RawTracepointWritable (0.00s)
        prog_test.go:76: Program type RawTracepointWritable isn't supported even though kernel is at least 5.2: not supported
    --- FAIL: TestHaveProgType/RawTracepoint (0.00s)
        prog_test.go:76: Program type RawTracepoint isn't supported even though kernel is at least 4.17: not supported
    --- FAIL: TestHaveProgType/PerfEvent (0.00s)
        prog_test.go:76: Program type PerfEvent isn't supported even though kernel is at least 4.9: not supported
    --- FAIL: TestHaveProgType/Kprobe (0.00s)
        prog_test.go:76: Program type Kprobe isn't supported even though kernel is at least 4.1: not supported
    --- FAIL: TestHaveProgType/TracePoint (0.00s)
        prog_test.go:76: Program type TracePoint isn't supported even though kernel is at least 4.7: not supported
```

This matches bpftool output:
```
[timo@alarm-rpi4 ebpf]$ sudo bpftool feature probe | grep NOT
eBPF program_type kprobe is NOT available
eBPF program_type tracepoint is NOT available
eBPF program_type perf_event is NOT available
eBPF program_type raw_tracepoint is NOT available
eBPF program_type lirc_mode2 is NOT available
eBPF program_type raw_tracepoint_writable is NOT available
eBPF program_type cgroup_sockopt is NOT available
eBPF program_type tracing is NOT available
eBPF program_type ext is NOT available
eBPF program_type lsm is NOT available
eBPF map_type xskmap is NOT available
eBPF map_type struct_ops is NOT available
eBPF map_type inode_storage is NOT available
eBPF map_type task_storage is NOT available
```

It seems like, for now, the error reported here is expected. I've played around with arm32 on an older Pi before, and attempting to use kprobes/tracepoints there resulted in consistent kernel panics. I think it's just been disabled in recent builds, or the kernel support is just not ready.

I'll slowly investigate some of the other failures too, but they all look related to either u/kprobes, perf, or missing vmlinux BTF somehow.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -1,3 +1,4 @@
+name: ci
 on:
   push:
     branches: [ "main" ]
@@ -69,13 +70,6 @@ jobs:
           go build -v ./...
           go test -c -o /dev/null ./... >/dev/null
 
-      - name: Cross build arm64
-        env:
-          GOARCH: arm64
-        run: |
-          go build -v ./...
-          go test -c -o /dev/null ./... >/dev/null
-
   build-docs:
     name: Build Documentation
     runs-on: ubuntu-22.04
@@ -138,6 +132,39 @@ jobs:
         with:
           name: Test Results (previous stable Go)
           path: junit.xml
+
+  test-on-arm64:
+    name: Run tests on arm64
+    runs-on: actuated-arm64-2cpu-8gb
+    needs: build-and-lint
+    timeout-minutes: 10
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Set up Go
+        uses: actions/setup-go@v4
+        with:
+          go-version: '${{ env.prev_go_version }}'
+
+      - run: go install gotest.tools/gotestsum@v1.8.1
+
+      - name: Test
+        run: gotestsum --ignore-non-json-output-lines --junitfile junit.xml -- -exec sudo -short -count 1 -json ./...
+
+      - name: Benchmark
+        run: go test -exec sudo -short -run '^$' -bench . -benchtime=1x ./...
+
+      - name: Upload Test Results
+        if: always()
+        uses: actions/upload-artifact@v3
+        with:
+          name: Test Results (arm64)
+          path: junit.xml
+
+      - name: Show dmesg
+        if: failure()
+        run: |
+          sudo dmesg
 
   vm-test:
     name: Run tests on pre-built kernel

--- a/link/uprobe_test.go
+++ b/link/uprobe_test.go
@@ -71,15 +71,17 @@ func TestExecutableLazyLoadSymbols(t *testing.T) {
 	ex, err := OpenExecutable("/bin/bash")
 	qt.Assert(t, qt.IsNil(err))
 	// Addresses must be empty, will be lazy loaded.
-	qt.Assert(t, qt.DeepEquals(ex.addresses, map[string]uint64{}))
+	qt.Assert(t, qt.HasLen(ex.addresses, 0))
 
 	prog := mustLoadProgram(t, ebpf.Kprobe, 0, "")
-	up, err := ex.Uprobe(bashSym, prog, &UprobeOptions{Address: 123})
+	// Address must be a multiple of 4 on arm64, see
+	// https://elixir.bootlin.com/linux/v6.6.4/source/arch/arm64/kernel/probes/uprobes.c#L42
+	up, err := ex.Uprobe(bashSym, prog, &UprobeOptions{Address: 124})
 	qt.Assert(t, qt.IsNil(err))
 	up.Close()
 
 	// Addresses must still be empty as Address has been provided via options.
-	qt.Assert(t, qt.DeepEquals(ex.addresses, map[string]uint64{}))
+	qt.Assert(t, qt.HasLen(ex.addresses, 0))
 
 	up, err = ex.Uprobe(bashSym, prog, nil)
 	qt.Assert(t, qt.IsNil(err))

